dlog/plonk/src/index.rs

```diff
@@ -0,0 +1,100 @@
+/*****************************************************************************************************************
+
+This source file implements Plonk Protocol Index primitive.
+
+*****************************************************************************************************************/
+
+use commitment_dlog::{srs::SRS, commitment::{CommitmentCurve, PolyComm}};
+use ff_fft::EvaluationDomain;
+use algebra::AffineCurve;
+use oracle::rndoracle::ProofError;
+use oracle::poseidon::ArithmeticSpongeParams;
+use plonk_circuits::{gate::CircuitGate, constraints::ConstraintSystem};
+use algebra::PrimeField;
+
+type Fr<G> = <G as AffineCurve>::ScalarField;
+type Fq<G> = <G as AffineCurve>::BaseField;
+
+pub enum SRSValue<'a, G : CommitmentCurve> {
+    Value(SRS<G>),
+    Ref(&'a SRS<G>)
+}
+
+impl<'a, G : CommitmentCurve> SRSValue<'a, G> {
+    pub fn get_ref(&self) -> & SRS<G> {
+        match self {
+            SRSValue::Value(x) => &x,
+            SRSValue::Ref(x) => x
+        }
+    }
+}
+
+pub enum SRSSpec <'a, G: CommitmentCurve>{
+    Use(&'a SRS<G>),
+    Generate
+}
+
+impl<'a, G: CommitmentCurve> SRSValue<'a, G> where G::BaseField : PrimeField {
+    pub fn generate(size: usize) -> SRS<G> {
+        SRS::<G>::create(size)
+    }
+
+    pub fn create<'b>(size: usize, spec : SRSSpec<'a, G>) -> SRSValue<'a, G>{
+        match spec {
+            SRSSpec::Use(x) => SRSValue::Ref(x),
+            SRSSpec::Generate => SRSValue::Value(Self::generate(size))
+        }
+    }
+}
+
+pub struct Index<'a, G: CommitmentCurve>
```

Comment:
Nice, happy we could reuse our infrastructure!

Reply:
Good to hear.